### DISCUSSION TITLE
How to define a job from a list of asset and an op

### DISCUSSION THREAD

**Discussion by ShootingStarD** [dagster-io/dagster#19372] [Q&A]
Hi,
I want to define a Dagster job. This job should launch a list of Asset but the last step of our job should be to push to an external party the final results of our pipeline, so for me it should be an op , not an asset, as assets should return something. This is the equivalent to the [sending a mail example in the doc ](https://docs.dagster.io/guides/dagster/how-assets-relate-to-ops-and-graphs#op-graphs-that-read-from-an-asset)
I can easily define a job based on assets of a module : 
```python
my_job = define_asset_job(
    name="my_job",
    selection=AssetSelection.assets(
        asset1,
        asset2,
    ),
    partitions_def=five_minutes_partitions,
)
```

However I do not understand how to add my op because it is not an asset and therefore cannot be give to define_asset_job()
Do you have any idea on how to do this?

**Comment by the4thamigo-uk** [comment]
Well I don't know the answer, but assets don't have to return data. We routinely have assets that represent data stored in external data stores, and your final asset could represent the data sent to the external party. To do this you can define your asset returning `Output[None]`.

**Comment by marcilj** [comment]
Multiple things here.

1. You might have your reason, but if the `push` you are talking about is to `store` the data somewhere in an external party software, you can still use and asset based approach and create a new IO Manager that should be able to store and retrieve the stored data.
If you just want to emit an event then an Ops make sense.
2. You should look into `graph` and `graph_assets` https://docs.dagster.io/guides/dagster/how-assets-relate-to-ops-and-graphs
https://docs.dagster.io/guides/dagster/how-assets-relate-to-ops-and-graphs#op-graphs-that-read-from-an-asset

Since you linked the same page as me I expect that you've already read it.

There are other ways to create jobs. In your example you use the function `define_asset_job` and that is only for assets.

If you want more control, you should build a graph then use the job decorator to create a job
```python
@job
def send_emails_job():
    send_emails(emails_to_send.to_source_asset())
```
